File: app/src/main/java/com/junerver/cloudnote/ui/activity/NoteDetailActivity.java
package com.junerver.cloudnote.ui.activity;

import android.content.DialogInterface;
import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.media.ThumbnailUtils;
import android.os.Bundle;
import android.provider.MediaStore;
import android.support.v7.app.AlertDialog;
import android.view.View;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.junerver.cloudnote.CloudNoteApp;
import com.junerver.cloudnote.R;
import com.junerver.cloudnote.db.entity.Note;
import com.junerver.cloudnote.db.entity.NoteEntity;
import com.orhanobut.logger.Logger;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import cn.bmob.v3.exception.BmobException;
import cn.bmob.v3.listener.UpdateListener;

/**
 * 用于查看笔记的页面
 */

public class NoteDetailActivity extends BaseActivity {


    @BindView(R.id.tvNoteTitle)
    TextView mTvNoteTitle;
    @BindView(R.id.tvNoteContent)
    TextView mTvNoteContent;
    @BindView(R.id.btnEdit)
    LinearLayout mBtnEdit;
    @BindView(R.id.btnDelete)
    LinearLayout mBtnDelete;
    @BindView(R.id.ivBack)
    ImageView mIvBack;
    @BindView(R.id.ivDone)
    ImageView mIvDone;
    @BindView(R.id.ibImage)
    ImageButton mIbImage;
    @BindView(R.id.ibVideo)
    ImageButton mIbVideo;

    private NoteEntity mNoteEntity;

    @Override
    protected void initView() {

        mIvDone.setVisibility(View.GONE);

        String title = mNoteEntity.getTitle();
        String content = mNoteEntity.getContent();
        String image = mNoteEntity.getImage();
        String video = mNoteEntity.getVideo();

        if (title != null) {
            mTvNoteTitle.setText(title);
        }
        if (content != null) {
            mTvNoteContent.setText(content);
        }
        if (image != null) {
            mIbImage.setImageBitmap(BitmapFactory.decodeFile(image));
        }
        if (video != null) {
            Bitmap bitmap = ThumbnailUtils.createVideoThumbnail(video, MediaStore.Images.Thumbnails.MICRO_KIND);
            bitmap = ThumbnailUtils.extractThumbnail(bitmap, Math.max(500, mIbVideo.getWidth()), Math.max(500, mIbVideo.getHeight()), ThumbnailUtils.OPTIONS_RECYCLE_INPUT);
            mIbVideo.setImageBitmap(bitmap);
        }

    }

    @Override
    protected void initData() {
        mNoteEntity = getIntent().getParcelableExtra("Note");
    }

    @Override
    protected void setListeners() {

    }

    @Override
    protected int getLayoutId() {
        return R.layout.activity_note_detail;
    }


    @OnClick({R.id.btnEdit, R.id.btnDelete, R.id.ivBack, R.id.ibImage, R.id.ibVideo})
    public void onClick(View view) {
        switch (view.getId()) {
            case R.id.btnEdit:
                Intent editIntent = new Intent(mContext, EditNoteActivity.class);
                editIntent.putExtra("Note", mNoteEntity);
                startActivity(editIntent);
                break;
            case R.id.btnDelete:
                AlertDialog.Builder builder = new AlertDialog.Builder(mContext);
                builder.setTitle("确认要删除这个笔记么？");
                builder.setPositiveButton("确认", new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialogInterface, int i) {
                        Note note = new Note();
                        note.setObjectId(mNoteEntity.getObjId());
                        note.delete(new UpdateListener() {
                            @Override
                            public void done(BmobException e) {
                                if (e == null) {
                                    Logger.i("bmob数据删除成功");
                                    //清除本地数据
                                    CloudNoteApp.getNoteEntityDao().delete(mNoteEntity);
                                    showShortToast("笔记已删除！");
                                    finish();
                                } else {
                                    Logger.i("bmob数据删除失败：" + e.getMessage() + "," + e.getErrorCode());
                                    showShortToast("错误！请稍后再试！");
                                }
                            }
                        });
                    }
                });
                builder.setNegativeButton("取消", null);
                builder.show();
                break;
            case R.id.ivBack:
                finish();
                break;
            case R.id.ibImage:
                //查看图片
                break;
            case R.id.ibVideo:
                //查看视频
                break;
        }
    }


}

File: app/src/main/java/com/junerver/cloudnote/ui/activity/NoteDetailActivity.kt
package com.junerver.cloudnote.ui.activity

import android.app.Activity
import android.content.Intent
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.media.ThumbnailUtils
import android.provider.MediaStore
import android.view.View
import androidx.appcompat.app.AlertDialog
import androidx.lifecycle.lifecycleScope
import com.junerver.cloudnote.net.BmobMethods
import com.edusoa.ideallecturer.fetchNetwork
import com.edusoa.ideallecturer.toBean
import com.elvishew.xlog.XLog
import com.junerver.cloudnote.R
import com.junerver.cloudnote.bean.DelResp
import com.junerver.cloudnote.bean.ErrorResp
import com.junerver.cloudnote.databinding.ActivityNoteDetailBinding
import com.junerver.cloudnote.databinding.BackBarBinding
import com.junerver.cloudnote.db.NoteUtils
import com.junerver.cloudnote.db.entity.NoteEntity
import com.junerver.cloudnote.utils.NetUtils
import kotlinx.coroutines.launch

/**
 * 用于查看笔记的页面
 */
class NoteDetailActivity : BaseActivity<ActivityNoteDetailBinding>() {

    //看起来是个entity，其实不是
    private lateinit var mNote: NoteEntity

    private lateinit var backBarBinding: BackBarBinding
    override fun initData() {
        mNote = intent.getSerializableExtra("Note")!! as NoteEntity
        XLog.d(mNote)
    }

    override fun initView() {
        backBarBinding = BackBarBinding.bind(viewBinding.llRoot)
        backBarBinding.ivDone.visibility = View.GONE
        backBarBinding.tvBarTitle.text = "笔记详情"
        inflateView(mNote)
    }

    private fun inflateView(noteEntity: NoteEntity) {
        val title: String = noteEntity.title
        val content: String = noteEntity.content
        val image: String = noteEntity.image
        val video: String = noteEntity.video
        if (title.isNotEmpty()) {
            viewBinding.tvNoteTitle.text = title
        }
        if (content.isNotEmpty()) {
            viewBinding.tvNoteContent.text = content
        }
        if (image.isNullOrEmpty()) {
            var bitmap: Bitmap = BitmapFactory.decodeFile(image)
            bitmap = ThumbnailUtils.extractThumbnail(
                bitmap,
                Math.max(500, viewBinding.ivImage.width),
                Math.max(500, viewBinding.ivImage.height),
                ThumbnailUtils.OPTIONS_RECYCLE_INPUT
            )
            viewBinding.ivImage.setImageBitmap(bitmap)
        }
        if (video.isNotEmpty()) {
            var bitmap: Bitmap = ThumbnailUtils.createVideoThumbnail(
                video,
                MediaStore.Images.Thumbnails.MICRO_KIND
            )!!
            bitmap = ThumbnailUtils.extractThumbnail(
                bitmap,
                Math.max(500, viewBinding.ibVideo.width),
                Math.max(500, viewBinding.ibVideo.height),
                ThumbnailUtils.OPTIONS_RECYCLE_INPUT
            )
            viewBinding.ibVideo.setImageBitmap(bitmap)
        }
    }

    override fun setListeners() {
        viewBinding.btnEdit.setOnClickListener {
            val editIntent = Intent(mContext, EditNoteActivity::class.java)
            editIntent.putExtra("Note", mNote)
            startActivityForResult(editIntent, EDIT_NOTE)
        }
        viewBinding.btnDelete.setOnClickListener {
            val builder: AlertDialog.Builder = AlertDialog.Builder(mContext)
            builder.setTitle("确认要删除这个笔记么？")
            builder.setPositiveButton(
                "确认"
            ) { _, _ ->
                if (NetUtils.isConnected(mContext)) {
                    lifecycleScope.launch {
                        //云端删除
                        fetchNetwork {
                            doNetwork {
                                BmobMethods.INSTANCE.delNoteById(mNote.objId)
                            }
                            onSuccess {
                                val delResp = it.toBean<DelResp>()
                                //本地删除
                                NoteUtils.delNoteById(mNote.objId)
                                showShortToast(getString(R.string.del_success))
                                finish()
                            }
                            onHttpError { errorBody, errorMsg, code ->
                                errorBody?.let {
                                    val bean = it.toBean<ErrorResp>()
                                    delInLocalWithoutNetWork()
                                }
                            }
                        }
                    }
                } else {
                    delInLocalWithoutNetWork()
                }
            }
            builder.setNegativeButton("取消", null)
            builder.show()
        }
        backBarBinding.ivBack.setOnClickListener { finish() }
    }

    private fun delInLocalWithoutNetWork() {
        mNote.isLocalDel = true
        mNote.saveOrUpdate("objId = ?", mNote.objId)
        showShortToast(getString(R.string.del_success))
        finish()
    }

    //编辑后返回变更
    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (resultCode == Activity.RESULT_OK) {
            mNote = data?.getSerializableExtra("Note")!! as NoteEntity
            inflateView(mNote)
        }
    }

    companion object {
        private const val EDIT_NOTE = 3
    }
}